### ISSUE TITLE
MariaDB 10.4 needs update from 10.4.25 to 10.4.28 please for bug fixes. thank you!

### ISSUE THREAD

**Issue by EdNett** [omega8cc/boa#1699]
Hello,  We got stuck on MariaDB 10.4 a few years ago (whnen Boa supported 10.3 and 10.4) before you switched to Percona.

Now we find we have MariaDB 10.4.25 and this needs to be upgraded to 10.4.28! Can you make this version change in the baracuda up-head script, please so that we can stay secure and get the latest bug fixes.

We have an issue 
https://www.drupal.org/project/drupal/issues/3349244 
and an inability to create a new content page or article due to a invalid datetime format 1366 error since updating to 7.94 or 7.95 - not sure when the error began.

If you can't or won't do this, please let us know, and give advice as to whether we can update mariadb ourselves by changing the sources list and using aptitude (or could that cause the boa system to fail?)

or should we try to downgrade to 10.3 (are you maintaining upgrades in the maria 10.3 veriosn?) or try a hot switch to percona - and if so, any guidelines?

Thanks, we probably aren't alone in being stuck with Maria 10.4 and we do need to apply bug and security fixes, one wqay or another.

Thanks,

Ed

**Comment by reswild** [comment]
What distro are you running? If you are still on Debian Stretch, I don't think you get new versions for MariaDB in the regular repos, as Stretch is no longer supported. If you are able to safely upgrade to Buster, that should give you MariaDB 10.4.28.

**Comment by EdNett** [comment]
Thanks for your advice. I am going to wait for boa to have a new stable or semi-stable release, work on which appears to be happening, since I fear that mariadb won't be supported at all anymore, so I'll be forced to switch to Percona, and if that's the case, better ot to do any upgrading at present, I feel. Have yo done the stretch to buster upgrade with BOA - on your own, of with the _STRETCH_TO_BUSTER line flag when running a barracuda up-head system?

**Comment by EdNett** [comment]
I'd like to do a stretch to buster update, then go on to bullseye, however, I don't know what will happen with MariaDB 10.4 which we're on, unfortunately, so I am waiting to get a response from BOA maintainers about best way forward, whether or not maria is supported at all, etc.  Boa had a db update script a few years ago (I forget i which direction) and then there is the db script which updates collation to utf-8mb4 - perhaps one of those cuold be sued as a starting point and modified accordingly. If there is no easy way to do a maria 10.4 to percona 5.7 update, then I might make db exports for each site manually, save all /files/ directories recursively for each site and then reinstall and upload or rsync and copy all /files/*  directories for each site and import db with previously saved db exports, to make the upgrade.

**Comment by reswild** [comment]
Buster ought to work fine with MariaDB 10.4. I think Bullseye requires 10.5, so you probably want to do the MariaDB to Percona migration before your Buster to Bullseye upgrade.

The BOA update script only supports MariaDB 10.2 to Percona 5.7. I don't think there is an easy way to update existing databases from either 10.3 or 10.4, but it should be possible to swap databases without having to reinstall the sites or platforms.

**Comment by EdNett2** [comment]
Hi, It's me - I lost my github .biz domain name - too expensive to renew, and bought the .us version instead. Then I cleared my cookies off my browser, and can't receive the github verification email! So, I don't know if they will give me access to my old EdNett account.

In any case, what about downgrading mariadb 10.4 to 10.2 and then use the update script boa provides? I know mariadb doesn't officially support downgrading, but I have successfully done it in the past - even though it was a long time ago.  Would that be a better path forward?

Best,

Ed

**Comment by omega8cc** [comment]
We have explained this situation in the docs: https://github.com/omega8cc/boa/blob/master/docs/FAQ.txt#L2

**Comment by EdNett** [comment]
@omega8cc and @reswild 

Do you think this is a good procedure for mariadb downgrade 10.4 to 10.2: (and then run the mariadb-10.2 to percona update script that BOA has)?

mysqldump -u root -p --all-databases > databasedump.sql

[make individual db dumps for every db for added security]

nano /etc/apt/sources.list.d/mariadb.list  (change 10.4 to 10.2 in both lines)
apt install dirmngr
sudo apt-key adv --recv-keys --keyserver hkp://keyserver.ubuntu.com:80 0xF1656F24C74CD1D8


apt update
apt clean
apt autoclean
apt update

apt list installed mariadb-client mariadb-common mariadb-compat mariadb-devel mariadb-server mariadb-shared

apt remove mariadb-client mariadb-common mariadb-server  (and any other mariadb package which is installed)

mv /var/lib/mysql /tmp

apt clean
apt update

apt install mariadb-client mariadb-common mariadb-server

service mysql start
mysql_secure_installation
mysql -v

mysql -u root -p root < databasedump.sql

service mysql restart

Best, Ed

**Comment by reswild** [comment]
The updated docs say you can upgrade to Percona 5.7 from MariaDB 10.3, but I think this should say 10.2. Also, if you have to do a manual migration anyway, I would just migrate directly to Percona 5.7.

I haven't gotten around to testing any database migration scripts yet, but I would suggest to not include the mysql database in the dump and restore, but export/import users and privileges in a separate step. 

I hope to get around to do my own database migration later this month, so I can share my migration script then.

**Comment by omega8cc** [comment]
There is no point in running backups manually, just use:

`$ bash /var/xdrago/mysql_backup.sh` (you can edit this script to disable gzipping since you don't need it)

If you prefer to have classic (but much slower) single dump per database, move away mydumper temporarily :

`$ mv /usr/local/bin/mydumper /usr/local/bin/mydumper_off`

Also, don't backup and import the mysql main database as-is -- it may be a tricky point which will break things for you.

As for the compatible MariaDB version, we have moved all servers from 10.3 to Percona 5.7 -- although we didn't monitor what changes have been introduced in later 10.3.x releases.

Whatever you decide to test, don't do that on the production server blindly!

Create a clone from its backup so you could experiment and restore it from backup if needed.

**Comment by EdNett** [comment]
Well, I am on a bad (thru mobile phone connection) isp - and at the exact moment when I had removed 10.4 I lost my ssh connection. I have only been able to connect to the server in rescue mode since, and no matter hat I do, I cannot get an ssh connection. I have checked the csf allowed and ignored files, and added my IP to them, but I don't know what could be causing this problem.  I may have to rebuild entirely the production server unlesss I can get ssh connection back. When I login in rescur moos and mount the disks and chrrot etc. I see that ssh isn't running, and when I try to start it I get this error:
/dev/null is not a character device!

and it doesn't start on reboot either (soft or hard)

Any ideas on how to get sshd running again?  Thanks for advice. I see that csf is still runnin gbut lfd isn't and won't start (from rescue mode).

Lastly, can I hope to import a mariadb db into a percona system if I reinstall from scratch, or is that not goign to work?

Best,

Ed

**Comment by reswild** [comment]
As mentioned above, special care needs to be taken regarding the mysql database. Other databases should be OK to restore from a database dump, although this might depend on using the right format for the dump, so this should also be tested before you try this on a production server.

For your ssh issue, I guess you could try the suggestion from here: https://mithunkumr.wordpress.com/2012/10/17/unable-to-start-ssh-devnull-is-not-a-character-device/

Hopefully you made a snapshot or other type of backup before starting all this, so you can restore your system to a previous state if you can't get things back up and running again.

**Comment by reswild** [comment]
I've had problems in the past when trying to use a Hetzner snapshot to create a new server, but they worked as they should if I created the server first, and then rebuilt it using the snapshot.
